Support rtwdog on rt1180

SuperHeroAbner opened the pull request:
Added the rtwdog Zephyr driver for RT1180 EVK.

SuperHeroAbner committed:
drivers: rtwdog: add NXP rtwdog driver

Port NXP rtwdog driver to Zephyr.

Signed-off-by: Ruijia Wang <ruijia.wang@nxp.com>

SuperHeroAbner committed:
mcux: soc: rt1180 unmask reset event when rtwdog is using

RT1180 takes reset event mask feature which should be unmasked when
watch dog is enabled.

Signed-off-by: Ruijia Wang <ruijia.wang@nxp.com>

SuperHeroAbner commented:
Related hal_nxp PR: https://github.com/zephyrproject-rtos/hal_nxp/pull/482

lucien-nxp commented:
> Related hal_nxp PR: [zephyrproject-rtos/hal_nxp#482](https://github.com/zephyrproject-rtos/hal_nxp/pull/482)

Please record your hal_nxp PR in west.yml like : `pull/444/head`

pdgendt commented on drivers/watchdog/wdt_mcux_imx_rtwdog.c:
```diff
@@ -0,0 +1,222 @@
+	return 0;
+}
+
+static const struct wdt_driver_api mcux_rtwdog_api = {
```
Drivers need to wrap APIs.

```suggestion
static DEVICE_API(wdt, mcux_rtwdog_api) = {
```

danieldegrasse commented on drivers/watchdog/wdt_mcux_imx_rtwdog.c:
```diff
@@ -0,0 +1,222 @@
+
+static struct mcux_rtwdog_data mcux_rtwdog_data;
+
+DEVICE_DT_INST_DEFINE(0, &mcux_rtwdog_init, NULL, &mcux_rtwdog_data, &mcux_rtwdog_config,
```
Let's not use this syntax here, this driver can easily be multi instance. Take a look at this driver for an example of a multi-instance watchdog setup: https://github.com/zephyrproject-rtos/zephyr/blob/537b52e7fe8f183fc7e51f5d0948464cf71ffa38/drivers/watchdog/wdt_gecko.c#L281

danieldegrasse commented on dts/bindings/watchdog/nxp,imx-rtwdog.yaml:
```diff
@@ -0,0 +1,28 @@
+  clk-divider:
+    type: int
+    description: Watchdog counter clock divider
+    required: true
```
Missing newline at end of file

danieldegrasse commented on soc/nxp/imxrt/imxrt118x/soc.c:
```diff
@@ -56,6 +59,14 @@ static const clock_arm_pll_config_t armPllConfig_BOARD_BootClockRUN = {
 
+#if defined(CONFIG_WDT_MCUX_IMX_RTWDOG)
+#define RTWDOG_IF_SET_SRC(n, i)                                                                    \
+        if (strcmp(DT_PROP(DT_NODELABEL(rtwdog##n), status), "okay") == 0) {                       \
```
Use `if (IS_ENABLED(DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(rtwdog#nn), nxp_imx_rtwdog, okay))` here, instead of the runtime check.

SuperHeroAbner replied:
Updated. Thanks Daniel for all these great advises.

danieldegrasse reviewed (commented):
Please take a look at the compliance errors, the github review UI highlights the issues that need fixing nicely

SuperHeroAbner committed:
manifest: hal_nxp: update NXP HAL to include support for RTWDOG driver

Update the NXP HAL revision to include support for building the RTWDOG
driver.

Signed-off-by: Ruijia Wang <ruijia.wang@nxp.com>

SuperHeroAbner commented on dts/bindings/watchdog/nxp,imx-rtwdog.yaml:
```diff
@@ -0,0 +1,28 @@
+
+description: imxRT watchdog
+
+compatible: "nxp,imx-rtwdog"
```
Removed the key word like imx from all names.

hakehuang commented:
please add `-watchdog` to boards\nxp\mimxrt1180_evk\mimxrt1180_evk_mimxrt1189_cm33.yaml

danieldegrasse commented on west.yml:
Commit with the HAL changes needs to  be first in the series in order to make everything bisectable, so that the first commit adding the RTWDOG driver will actually build the driver. Also, please update the commit message here- something like the following would be better:

```
manifest: hal_nxp: update NXP HAL to include support for RTWDOG driver

Update the NXP HAL revision to include support for building the RTWDOG driver
```

danieldegrasse commented on boards/nxp/mimxrt1180_evk/mimxrt1180_evk.dtsi:
```diff
@@ -12,6 +12,7 @@
 		led0 = &green_led;
 		sw0 = &user_button;
 		pwm-led0 = &green_pwm_led;
+		watchdog0 = &rtwdog0;
```
Set this at the SOC level devicetree, since the watchdog is internal to the SOC. Example: https://github.com/zephyrproject-rtos/zephyr/blob/e60f04096cd8d9227c8279ca830e72bc1b5b8b1c/dts/arm/nxp/nxp_mcxw71.dtsi#L17

danieldegrasse commented on drivers/watchdog/wdt_mcux_rtwdog.c:
```diff
@@ -0,0 +1,188 @@
+			      &mcux_rtwdog_config_##n, POST_KERNEL,                                \
+			      CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &mcux_rtwdog_api);
+
+DT_INST_FOREACH_STATUS_OKAY(MCUX_RTWDOG_INIT)
```
Missing newline here. You may want to check your editor settings to make sure it adds a newline at the end of the file (some editors can also strip trailing whitespace)

danieldegrasse commented:
> please add `-watchdog` to boards\nxp\mimxrt1180_evk\mimxrt1180_evk_mimxrt1189_cm33.yaml

Also would be good to document support for this on the board, like we do here: https://github.com/zephyrproject-rtos/zephyr/blob/e60f04096cd8d9227c8279ca830e72bc1b5b8b1c/boards/nxp/rd_rw612_bga/doc/index.rst?plain=1#L53

danieldegrasse commented:
Please rebase when you have a chance- sorry that the manifest file keeps getting conflicts!

dleach02 commented:
@SuperHeroAbner please rebase so we can approve and merge.

SuperHeroAbner commented:
> @SuperHeroAbner please rebase so we can approve and merge.

Done. Should we merge [this PR](https://github.com/zephyrproject-rtos/hal_nxp/pull/482) first? Then I delete the PR link commit in current PR, then merge this PR.

dleach02 commented:
@SuperHeroAbner we need to have this PR clean on CI and with approvals ready. 

The sequence is when the Zephyr side is ready we will merge the HAL and then update the west.yml in this PR to point to the new SHA. We can't do that until the Zephyr side PR is ready.

mmahadevan108 commented:
@SuperHeroAbner I have merged the HAL side PR. Please help update the SHA

dleach02 commented:
@mmahadevan108 I updated @SuperHeroAbner PR
